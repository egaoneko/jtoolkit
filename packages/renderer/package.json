{
  "name": "@toolkit/renderer",
  "version": "0.0.1",
  "description": "renderer",
  "main": "lib/index",
  "types": "lib/index",
  "files": [
    "lib"
  ],
  "license": "MIT",
  "scripts": {
    "build": "npm run compile",
    "compile": "tsc --project tsconfig.json",
    "start": "npm run build && node lib/index.js",
    "test": "jest",
    "test:watch": "jest --watch",
    "lint": "eslint \"src/**/*.{js,ts}\" --fix",
    "format": "prettier --write \"src/**/*.ts\""
  },
  "dependencies": {},
  "devDependencies": {}
}