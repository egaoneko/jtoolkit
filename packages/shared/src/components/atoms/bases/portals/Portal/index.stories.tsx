import { Story } from '@storybook/react';

import Portal, { PortalProps } from './index';
import { useEffect, useRef, useState } from 'react';

export default {
  title: 'atoms/bases/portal/Portal',
  component: Portal,
};

const BodyTemplate: Story<PortalProps> = () => {
  return (
    <Portal>
      <span>Portal to Body</span>
    </Portal>
  );
};

export const Body = BodyTemplate.bind({});

const ContainerTemplate: Story<PortalProps> = () => {
  const ref = useRef<HTMLDivElement>(null);
  const [mounted, setMounted] = useState<boolean>(false);
  useEffect(() => {
    setMounted(!!ref.current);
  }, [ref]);
  return (
    <>
      <div className="flex flex-col gap-2">
        Target
        <div className="w-[200px] border border-red-500" ref={ref} />
      </div>
      {mounted && (
        <Portal container={ref.current}>
          <span>Portal to Container</span>
        </Portal>
      )}
    </>
  );
};

export const Container = ContainerTemplate.bind({});
