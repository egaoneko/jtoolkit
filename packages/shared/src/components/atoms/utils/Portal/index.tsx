import { cloneElement, forwardRef, isValidElement, ReactElement, ReactNode, RefAttributes, useState } from 'react';
import ReactDOM from 'react-dom';
import { setRef, useEnhancedEffect, useForkRef } from '@toolkit/util';

export interface PortalProps {
  children?: ReactNode;
  container?: Element | (() => Element | null) | null;
  disablePortal?: boolean;
}

const Portal = forwardRef<Element, PortalProps>(function Portal(
  { children, container, disablePortal = false },
  ref,
): ReactElement | null {
  const [mountNode, setMountNode] = useState<Element | null>(null);
  const handleRef = useForkRef(isValidElement(children) ? (children as RefAttributes<Element>).ref : null, ref);

  useEnhancedEffect(() => {
    if (!disablePortal) {
      setMountNode(getContainer(container) || document.body);
    }
  }, [container, disablePortal]);

  useEnhancedEffect(() => {
    if (mountNode && !disablePortal) {
      setRef(ref, mountNode);
      return () => {
        setRef(ref, null);
      };
    }

    return undefined;
  }, [ref, mountNode, disablePortal]);

  if (disablePortal) {
    if (isValidElement(children)) {
      return cloneElement(children, {
        ref: handleRef,
      });
    }
    return children as ReactElement | null;
  }

  return mountNode ? ReactDOM.createPortal(children, mountNode) : mountNode;
});

export default Portal;

function getContainer(container?: Element | (() => Element | null) | null): Element | null {
  return typeof container === 'function' ? container() : container ?? null;
}
