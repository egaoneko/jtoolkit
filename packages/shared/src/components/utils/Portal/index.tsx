import { cloneElement, forwardRef, isValidElement, ReactElement, ReactNode, RefAttributes, useState } from 'react';
import ReactDOM from 'react-dom';
import { useIsomorphicLayoutEffect } from 'react-use';
import { setRef, useForkRef } from '@toolkit/util';

export interface PortalProps {
  container?: Element | (() => Element | null) | null;
  disablePortal?: boolean;
  children?: ReactNode;
}

const Portal = forwardRef<Element, PortalProps>(function Portal(
  { children, container, disablePortal = false },
  ref,
): ReactElement | null {
  const [mountNode, setMountNode] = useState<Element | null>(null);
  const handleRef = useForkRef(isValidElement(children) ? (children as RefAttributes<Element>).ref : null, ref);

  useIsomorphicLayoutEffect(() => {
    if (!disablePortal) {
      setMountNode(getContainer(container) || document.body);
    }
  }, [container, disablePortal]);

  useIsomorphicLayoutEffect(() => {
    if (mountNode && !disablePortal) {
      setRef(ref, mountNode);
      return () => {
        setRef(ref, null);
      };
    }

    return undefined;
  }, [ref, mountNode, disablePortal]);

  if (disablePortal) {
    if (isValidElement(children)) {
      return cloneElement(children, {
        ref: handleRef,
      });
    }
    return children as ReactElement | null;
  }

  return mountNode ? ReactDOM.createPortal(children, mountNode) : mountNode;
});

export default Portal;

function getContainer(container?: Element | (() => Element | null) | null): Element | null {
  return typeof container === 'function' ? container() : container ?? null;
}
