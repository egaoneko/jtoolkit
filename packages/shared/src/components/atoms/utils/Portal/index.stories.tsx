import { Story } from '@storybook/react';
import { useEffect, useRef, useState } from 'react';

import Portal, { PortalProps } from 'src/components/atoms/utils/Portal/index';
import styles from 'src/styles/storybook.css';

export default {
  title: 'atoms/utils/Portal',
  component: Portal,
};

const BodyTemplate: Story<PortalProps> = args => {
  return (
    <Portal {...args}>
      <span>Portal to Body</span>
    </Portal>
  );
};

export const Body = BodyTemplate.bind({});

const ContainerTemplate: Story<PortalProps> = args => {
  const ref = useRef<HTMLDivElement>(null);
  const [mounted, setMounted] = useState<boolean>(false);
  useEffect(() => {
    setMounted(!!ref.current);
  }, [ref]);
  return (
    <>
      <div className={styles.col}>
        Target
        <div className={styles.box} ref={ref} />
      </div>
      {mounted && (
        <Portal {...args} container={ref.current}>
          <span>Portal to Container</span>
        </Portal>
      )}
    </>
  );
};

export const Container = ContainerTemplate.bind({});
