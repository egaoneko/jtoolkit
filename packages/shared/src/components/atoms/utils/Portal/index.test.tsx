import renderer from 'react-test-renderer';
import ReactDOM from 'react-dom';
import { ReactNode, ReactPortal } from 'react';

import Portal from 'src/components/atoms/utils/Portal/index';

describe('Portal Snapshots', () => {
  let portalContainer: Element | null = null;
  beforeAll(() => {
    ReactDOM.createPortal = jest.fn((children: ReactNode, container: Element) => {
      portalContainer = container;
      return children;
    }) as (children: ReactNode, container: Element, key?: null | string) => ReactPortal;
  });

  afterEach(() => {
    portalContainer = null;
    (ReactDOM.createPortal as jest.Mock).mockClear();
  });

  test('renders portal with default options', async () => {
    const tree = renderer.create(
      <Portal>
        <div>Test</div>
      </Portal>,
    );
    await renderer.act(async () => {
      expect(tree.toJSON()).toMatchSnapshot();
      expect(portalContainer).toBe(document.body);
    });
  });

  test('renders portal with container', async () => {
    const container = document.createElement('div');
    const tree = renderer.create(
      <Portal container={container}>
        <div>Test</div>
      </Portal>,
    );
    await renderer.act(async () => {
      expect(tree.toJSON()).toMatchSnapshot();
      expect(tree.root.props.container).toBe(container);
      expect(portalContainer).toBe(container);
    });
  });

  test('renders portal with disablePortal', async () => {
    const tree = renderer.create(
      <Portal disablePortal>
        <div>Test</div>
      </Portal>,
    );
    await renderer.act(async () => {
      expect(tree.toJSON()).toMatchSnapshot();
      expect(tree.root.props.disablePortal).toBeTruthy();
      expect(portalContainer).toBeNull();
    });
  });
});
