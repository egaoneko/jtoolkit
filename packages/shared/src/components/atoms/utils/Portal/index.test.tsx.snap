// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Portal Snapshots renders portal with container 1`] = `
<div>
  Test
</div>
`;

exports[`Portal Snapshots renders portal with default options 1`] = `
<div>
  Test
</div>
`;

exports[`Portal Snapshots renders portal with disablePortal 1`] = `
<div>
  Test
</div>
`;
